src/debugger.rs

```diff
@@ -0,0 +1,218 @@
+use std::collections::HashMap;
+use std::io::Write;
+
+trait DebuggerCommand {
+    /// Handle the command and any inputs; returning true will exit the debugger
+    fn handle(
+        &self,
+        context: &mut Context,
+        cli: &DebuggerRepl,
+        matches: &ArgMatches,
+    ) -> Result<bool, String>;
+    fn about(&self) -> &'static str;
+}
+
+struct DebuggerRepl {
+    commands: HashMap<&'static str, Box<dyn DebuggerCommand>>,
+    output: RefCell<Box<dyn Write>>,
+}
+
+fn flush() -> Result<(), String> {
+    std::io::stdout().flush().map_err(|e| e.to_string())
+}
+
+impl DebuggerRepl {
+    fn new() -> Self {
+        DebuggerRepl {
+            commands: HashMap::new(),
+            output: RefCell::new(Box::new(std::io::stdout())),
+        }
+    }
+
+    fn register_command(&mut self, name: &'static str, handler: Box<dyn DebuggerCommand>) {
+        self.commands.insert(name, handler);
+    }
+
+    fn get_command(&self, name: &str) -> Option<&dyn DebuggerCommand> {
+        self.commands.get(name).map(|command| &**command)
+    }
+
+    fn writeln(&self, formatted_string: &str) -> Result<(), String> {
+        writeln!(self.output.borrow_mut(), "{formatted_string}").map_err(|e| e.to_string())?;
+        flush()?;
+        Ok(())
+    }
+
+    fn build_cli(&self) -> Command {
+        let mut cli = Command::new("repl")
+            .multicall(true)
+            .arg_required_else_help(true)
+            .subcommand_required(true)
+            .subcommand_value_name("DEBUGGER")
+            .subcommand_help_heading("IXA DEBUGGER")
+            .help_template("{all-args}");
+
+        for (name, handler) in &self.commands {
+            cli = cli.subcommand(Command::new(*name).about(handler.about()).help_template(
+                "{about-with-newline}\n{usage-heading}\n    {usage}\n\n{all-args}{after-help}",
+            ));
+        }
+
+        cli
+    }
+}
+
+/// Returns the current population of the simulation
+struct PopulationCommand;
+impl DebuggerCommand for PopulationCommand {
+    fn about(&self) -> &'static str {
+        "Get the total number of people"
+    }
+    fn handle(
+        &self,
+        context: &mut Context,
+        cli: &DebuggerRepl,
+        _matches: &ArgMatches,
+    ) -> Result<bool, String> {
+        cli.writeln(&format!("{}", context.get_current_population()))?;
+        Ok(false)
+    }
+}
+
+/// Exits the debugger and continues the simulation
+struct ContinueCommand;
+impl DebuggerCommand for ContinueCommand {
+    fn about(&self) -> &'static str {
+        "Continue the simulation and exit the debugger"
+    }
+    fn handle(
+        &self,
+        context: &mut Context,
+        cli: &DebuggerRepl,
+        _matches: &ArgMatches,
+    ) -> Result<bool, String> {
+        cli.writeln(&format!(
+            "Continuing the simulation from t = {}",
+            context.get_current_time()
+        ))?;
+        Ok(true)
+    }
+}
+
+fn readline() -> Result<String, String> {
```

Comment:
I think this is fine for now, but in the future I would use `rustylines`

Reply:
https://github.com/CDCgov/ixa/issues/126